src/librepgp/stream-sig.cpp

```diff
@@ -45,18 +45,51 @@
 
 #include <time.h>
 
-void
-signature_hash_key(const pgp_key_pkt_t &key, rnp::Hash &hash)
+/* computes the hash according to the key version (for fingerprint calculation) or signature
+ * version */
+static void
+hash_key(const pgp_key_pkt_t &key, rnp::Hash &hash, pgp_version_t key_or_sig_version)
 {
     if (!key.hashed_data) {
         /* call self recursively if hashed data is not filled, to overcome const restriction */
         pgp_key_pkt_t keycp(key, true);
         keycp.fill_hashed_data();
-        signature_hash_key(keycp, hash);
+        hash_key(keycp, hash, key_or_sig_version);
         return;
     }
 
-    switch (key.version) {
+    /* first check that we can actually fit the hashed length in the 2 or 4 byte fields */
+    switch (key_or_sig_version) {
+    case PGP_V2:
+        FALLTHROUGH_STATEMENT;
+    case PGP_V3:
+        FALLTHROUGH_STATEMENT;
+    case PGP_V4:
+        if (key.hashed_len > ((size_t) 1 << 16) - 1) {
```

Comment:
Given that we cannot have key artifacts with length >65535 at the moment (we will not parse it nor generate) we could add just an assert() here and below. The price of mistake would be just an incorrect signature.